components/script/dom/path2d.rs

```diff
@@ -59,24 +63,342 @@ impl Path2D {
+                    .skip(1)
+                    .position(|segment| *segment == PathSegment::ClosePath)
+                {
+                    Some(index) => index + 1,
+                    None => 0,
+                };
+
+                let subpath_first_point = path.get(subpath_first_point_index)?;
+                match subpath_first_point {
+                    PathSegment::MoveTo { x, y } |
+                    PathSegment::LineTo { x, y } |
+                    PathSegment::Quadratic { x, y, .. } |
+                    PathSegment::Bezier { x, y, .. } |
+                    PathSegment::Ellipse { x, y, .. } |
+                    PathSegment::SvgArc { x, y, .. } => Some(Point2D::new(*x, *y)),
+                    PathSegment::ClosePath => None,
+                    PathSegment::ArcTo {
+                        cp1x,
+                        cp1y,
+                        cp2x,
+                        cp2y,
+                        radius,
+                    } => {
+                        let cp0 = match Path2D::compute_final_arc_to_point(
+                            &path[..subpath_first_point_index],
+                        ) {
+                            Some(point) => point,
+                            None => Point2D::new(*cp1x, *cp1y),
+                        };
+                        let cp1 = Point2D::new(*cp1x, *cp1y);
+                        let cp2 = Point2D::new(*cp2x, *cp2y);
+                        Some(compute_arc_to_points(cp0, cp1, cp2, *radius))
+                    },
+                }
+            },
+            PathSegment::MoveTo { x, y } |
+            PathSegment::LineTo { x, y } |
+            PathSegment::Quadratic { x, y, .. } |
+            PathSegment::Bezier { x, y, .. } |
+            PathSegment::Ellipse { x, y, .. } |
+            PathSegment::SvgArc { x, y, .. } => Some(Point2D::new(*x, *y)),
+            PathSegment::ArcTo {
+                cp1x,
+                cp1y,
+                cp2x,
+                cp2y,
+                radius,
+            } => {
+                let last_segment = path.len() - 1;
+                let cp0 = match Path2D::compute_final_arc_to_point(&path[..last_segment]) {
+                    Some(point) => point,
+                    None => Point2D::new(*cp1x, *cp1y),
+                };
+                let cp1 = Point2D::new(*cp1x, *cp1y);
+                let cp2 = Point2D::new(*cp2x, *cp2y);
+                Some(compute_arc_to_points(cp0, cp1, cp2, *radius))
+            },
+        }
+    }
+
+    fn last_point(path: &[PathSegment]) -> Option<Point2D<f32>> {
+        let last_point = path.last()?;
+
+        match last_point {
+            PathSegment::MoveTo { x, y } |
+            PathSegment::LineTo { x, y } |
+            PathSegment::Quadratic { x, y, .. } |
+            PathSegment::Bezier { x, y, .. } |
+            PathSegment::Ellipse { x, y, .. } |
+            PathSegment::SvgArc { x, y, .. } => Some(Point2D::new(*x, *y)),
```

Comment:
shouldn't we use last xy pair here, given that we search for last not first point?

Reply:
My bad, I forgot how destruct worked in pattern matching, xy are always of endpoint: `PathSegment::Quadratic { cpx, cpy, x, y }` and all others are names control points.